### PR TITLE
Vm cleanup

### PR DESCRIPTION
**Pull request by bjourne**
Here is a small pr to make the cpp code cleaner. I made a macro `FACTOR_FOR_EACH` to make iteration nicer. I also added some functions to io.cpp so that if you run `factor -i=/tmp` it doesn't crash.

### CODE REVIEW

File: vm/utilities.hpp [hunk 1]
```diff
@@ -1,5 +1,11 @@
 namespace factor {
 
+// Poor mans range-based for loops.
+#define FACTOR_FOR_EACH(iterable)                           \
+  for (typeof((iterable).begin()) iter = (iterable).begin();    \
```
Comment (mrjbq7) [inline vm/utilities.hpp]: Does this call ``begin()`` twice?

File: vm/utilities.hpp [hunk 2]
```diff
@@ -1,5 +1,11 @@
...
+       iter != (iterable).end();                                \
```
Comment (mrjbq7) [inline vm/utilities.hpp]: And does this call ``end()`` on each loop?

File: vm/image.cpp [hunk 3]
```diff
@@ -253,7 +252,7 @@ void factor_vm::load_image(vm_parameters* p) {
   load_data_heap(file, &h, p);
   load_code_heap(file, &h, p);
 
-  safe_fclose(file);
```
Comment (mrjbq7) [inline vm/image.cpp]: Why change this one?  To hide ``io-error`` if close fails?

Comment (erg) [inline vm/image.cpp]: ``safe_fclose`` is wrong anyway -- you're not allowed to retry on ``EINTR``
on ``flclose`` or ``close``.

On Wed, Jun 3, 2015 at 11:24 AM, John Benediktsson <notifications@github.com

File: vm/io.cpp [hunk 4]
```diff
@@ -13,6 +13,39 @@ The Factor library provides platform-specific code for Unix and Windows
 with many more capabilities so these words are not usually used in
 normal operation. */
 
+int raw_fread(void* ptr, size_t size, size_t nitems, FILE* stream) {
+  size_t items_read = 0;
```
Comment (mrjbq7) [inline vm/io.cpp]: Seems weird to return a ``size_t`` cast to an ``int``.

File: vm/io.cpp [hunk 5]
```diff
@@ -13,6 +13,39 @@ The Factor library provides platform-specific code for Unix and Windows
...
+
+  do {
+    size_t ret = fread((void*)((int*)ptr + items_read * size), size,
+                       nitems - items_read, stream);
+    if (ret == 0) {
```
Comment (mrjbq7) [inline vm/io.cpp]: The man pages say:

| If an error occurs, or the end-of-file is reached, the return value is a short object count (or zero).

So perhaps you need to check ``if (ret <= expected)``.

Comment (bjourne) [inline vm/io.cpp]: Ok, I'll look into it closer. The idea was to return `-1` on error, therefore the size_t -> int change.

File: vm/io.cpp (hunk 4 above)
Comment (mrjbq7) [inline vm/io.cpp]: Maybe an ``ssize_t`` and overflow detection?  I dunno.

File: vm/image.cpp (hunk 3 above)
Comment (bjourne) [inline vm/image.cpp]: It's replaced with `raw_fclose`. Now if it fails, it will not segfault Factor.

**Comment by bjourne** [comment]
Updated the pr. The `FACTOR_FOR_EACH` is fixed so now end() is only called once. I also changed the raw_fread/safe_fread combo back to returning a size_t. It returns 0 on error.

**Comment by mrjbq7** [comment]
Looks pretty good to me, @erg?

**Comment by erg** [comment]
Looks good.

On Thu, Jun 4, 2015 at 11:05 PM, John Benediktsson <notifications@github.com

**Comment by mrjbq7** [comment]
Merged, thanks!!

**Pull request closed without merging**